heat/nn/data_parallel.py

```diff
@@ -0,0 +1,275 @@
+                layer_name_prev = layer_name
+                start_idx = idx
+
+            # register backward hooks for all model parameter tensors
+            if blocking:
+                param.register_hook(self._blocking_hook)
+            else:
+                param.register_hook(self._nonblocking_hook(layer_name, name))
+        self._param_slices[layer_name_prev] = slice(start_idx, len(self._param_indices))
+
+    def __setattr__(self, name, value):
+        # auto-detect end of epoch's training phase and finalize wait handles (only relevant for non-blocking)
+        if name == "training" and not value and not self.blocking:
+            self._async_update()
+        super(DataParallel, self).__setattr__(name, value)
+
+    def forward(self, *inputs: tuple, **kwargs: dict) -> torch.Tensor:
+        data = inputs[0]
+
+        if isinstance(data, ht.DNDarray):
+            lcl_data = data._DNDarray__array
+        elif isinstance(data, torch.Tensor):
+            lcl_data = data
+        else:
+            lcl_data = torch.tensor(data)
+
+        # check if non-blocking and training
+        if not self.blocking and self.module.training:
+            # register forward hooks for all layers with wait handles
+            for name, submodule in self.module.named_modules():
+                if name == "":
+                    continue
+                if name in self._layer_wait_handles:
+                    hook_handle = submodule.register_forward_pre_hook(self._forward_hook(name))
+                    self._fwd_hook_handles.append(hook_handle)
+        # perform forward pass
+        ret = self.module(lcl_data, *inputs[1:], **kwargs)
+
+        # finalize potentially remaining wait handles and update corresponding params (if
+        # computation graph has changed between previous backward and this forward)
+        if not self.blocking:
+            # set has to be copied in order to be manipulated during iteration
+            active_layers_cpy = self._active_layers.copy()
+            for layer_name in active_layers_cpy:
+                self._forward_hook(layer_name)(None, None)
+        # reset optimizer flag
+        self._update_next = False
+        # clear dictionary after all wait handles are used up (dynamic computation graph)
+        self._layer_wait_handles.clear()
+        # remove forward hooks (dynamic computation graph)
+        for hook_handle in self._fwd_hook_handles:
+            hook_handle.remove()
+        self._fwd_hook_handles.clear()
+
+        return ret
+
+    def update(self):
+        """
+        Force optimizer to update model parameters. For blocking, optimizer immediately updates parameters. For
+        non-blocking, optimizer will update parameters during next forward.
+        """
+        if self.blocking:
+            self.optimizer.step()
+        else:
+            self._update_next = True
+
+    def _async_update(self, param_slice: slice = None, layer_names: List[str] = None):
```

Comment:
The term "asynchronious"  needs an explanation, otherwise it may lead to confusion. In the literature, asynchronous SGD schemes (e.g.  J. Tsitsiklis, D. Bertsekas, and M. Athans. 1986.  
 Distributed asynchronous deterministic and stochastic gradient optimization algorithms.IEEE Trans.Automat. Control31, 9 (1986), 803–812.)  are methods in which the gradients from calculations of different time steps are averaged. The term has a different meaning here.

Reply:
ive changed it to `_iparam_update` to clear this up. thanks for pointing this out